### PR TITLE
Add support for cancel-able downloads

### PR DESCRIPTION
**Pull request by k-dominik**
Hi @FynnBe,

we talked about this a while ago. Main motivation is being able to support progress reporting other than in the cmd and immediate cancellation of long running downloads. All is done via a `pbar` that is passed through the functions until it reaches `_download_url`. We need this in ilastik 🐙

### CODE REVIEW

File: bioimageio/spec/shared/_resolve_source.py [hunk 1]
```diff
@@ -217,7 +222,18 @@ def resolve_rdf_source_and_type(
 
 
 @singledispatch  # todo: fix type annotations
-def resolve_source(source, root_path: typing.Union[os.PathLike, URI] = pathlib.Path(), output=None):
+def resolve_source(source, root_path: typing.Union[os.PathLike, URI] = pathlib.Path(), output=None, pbar=None):
+    """Resolve sources to local files
+
+    Args:
+        source: resource description as path, or uri
```
Comment (FynnBe) [inline bioimageio/spec/shared/_resolve_source.py:229]: a url to a weight file would not be a 'resource description' as in resource description file (RDF)... Of course it is not wrong to call it that, but I find this confusing
```suggestion
        source: e.g. a path or uri
```

File: bioimageio/spec/shared/_resolve_source.py [hunk 2]
```diff
@@ -217,7 +222,18 @@ def resolve_rdf_source_and_type(
...
+        root_path: path to model-rdf - used to resolve relative paths
+        output: file path to write contents to - if not given a file path is created
+        pbar: progress bar sharing a minimal tqdm interface, if none given, tqdm is used.
+          pbar is only used in the case of downloading resources. Specifying a custom pbar here
+          helps adding features like progress reporting (outside the cmd) and cancellation
+          (by raising CancelledError).
```
Comment (FynnBe) [inline bioimageio/spec/shared/_resolve_source.py:235]:
```suggestion
          (by raising DownloadCancelled).
```

File: bioimageio/spec/shared/_resolve_source.py [hunk 3]
```diff
@@ -240,7 +257,10 @@ def _resolve_source_uri_node(
 
 @resolve_source.register
 def _resolve_source_str(
-    source: str, root_path: typing.Union[os.PathLike, URI] = pathlib.Path(), output: typing.Optional[os.PathLike] = None
+    source: str,
+    root_path: typing.Union[os.PathLike, URI] = pathlib.Path(),
+    output: typing.Optional[os.PathLike] = None,
+    pbar=None,
 ) -> pathlib.Path:
     return resolve_source(fields.Union([fields.URI(), fields.Path()]).deserialize(source), root_path, output)
```
Comment (FynnBe) [inline bioimageio/spec/shared/_resolve_source.py:265]:
```suggestion
    return resolve_source(fields.Union([fields.URI(), fields.Path()]).deserialize(source), root_path, output, pbar)
```

File: bioimageio/spec/shared/_resolve_source.py [hunk 4]
```diff
@@ -31,6 +31,11 @@
 from .raw_nodes import URI
 
 
+class DownloadCancelled(Exception):
```
Comment (FynnBe) [inline bioimageio/spec/shared/_resolve_source.py:34]: as this is a private model, maybe we should move `DownloadCancelled` to `common`?

File: bioimageio/spec/shared/_resolve_source.py [hunk 5]
```diff
@@ -273,6 +294,7 @@ def _resolve_source_resolved_importable_path(
     source: raw_nodes.ResolvedImportableSourceFile,
     root_path: typing.Union[os.PathLike, URI] = pathlib.Path(),
     output: typing.Optional[os.PathLike] = None,
+    pbar=None,
 ) -> raw_nodes.ResolvedImportableSourceFile:
     return raw_nodes.ResolvedImportableSourceFile(
         callable_name=source.callable_name, source_file=resolve_source(source.source_file, root_path, output)
```
Comment (FynnBe) [inline bioimageio/spec/shared/_resolve_source.py:300]:
```suggestion
        callable_name=source.callable_name, source_file=resolve_source(source.source_file, root_path, output, pbar)
```

File: bioimageio/spec/shared/_resolve_source.py [hunk 6]
```diff
@@ -284,6 +306,7 @@ def _resolve_source_importable_path(
     source: raw_nodes.ImportableSourceFile,
     root_path: typing.Union[os.PathLike, URI] = pathlib.Path(),
     output: typing.Optional[os.PathLike] = None,
+    pbar=None,
 ) -> raw_nodes.ResolvedImportableSourceFile:
     return raw_nodes.ResolvedImportableSourceFile(
         callable_name=source.callable_name, source_file=resolve_source(source.source_file, root_path, output)
```
Comment (FynnBe) [inline bioimageio/spec/shared/_resolve_source.py:312]:
```suggestion
        callable_name=source.callable_name, source_file=resolve_source(source.source_file, root_path, output, pbar)
```

File: bioimageio/spec/shared/_resolve_source.py [hunk 7]
```diff
@@ -295,6 +318,7 @@ def _resolve_source_list(
     source: list,
     root_path: typing.Union[os.PathLike, URI] = pathlib.Path(),
     output: typing.Optional[typing.Sequence[typing.Optional[os.PathLike]]] = None,
+    pbar=None,
 ) -> typing.List[pathlib.Path]:
     assert output is None or len(output) == len(source)
     return [resolve_source(el, root_path, out) for el, out in zip(source, output or [None] * len(source))]
```
Comment (FynnBe) [inline bioimageio/spec/shared/_resolve_source.py:324]:
```suggestion
    assert pbar is None or len(pbar) == len(source)
    return [resolve_source(el, root_path, out, pb) for el, out, pb in zip(source, output or [None] * len(source), pbar)]
```

File: bioimageio/spec/shared/_resolve_source.py [hunk 8]
```diff
@@ -295,6 +318,7 @@ def _resolve_source_list(
     source: list,
     root_path: typing.Union[os.PathLike, URI] = pathlib.Path(),
     output: typing.Optional[typing.Sequence[typing.Optional[os.PathLike]]] = None,
+    pbar=None,
```
Comment (FynnBe) [inline bioimageio/spec/shared/_resolve_source.py:321]:
```suggestion
    pbar: typing.Sequence = None,
```

**Review by FynnBe — CHANGES_REQUESTED** [review]
pbar needs to be handed down in recursive `resolve_source` calls. otherwise this looks very good! 👍

**Comment by k-dominik** [comment]
thx @FynnBe for the review! I even found one more invocation of `resolve_source` that needed the `pbar` treatment.

File: bioimageio/spec/shared/_resolve_source.py (hunk 4 above)
Comment (FynnBe) [reply bioimageio/spec/shared/_resolve_source.py:34]: We can move later then when needed. I didn't see that it was available in shared.

**Pull request merged**